### PR TITLE
[WIP] MacOS inputcapture

### PR DESCRIPTION
**Pull request by meck**
I had a stab at writing a macos input capture module as i was tired not being able to cross the pointer back and forth between mac and linux, and it seemed a good rust exercise. There as some things left mainly modifier handling and updating the bounds on a monitor config change. But i wanted to put it up as a draft for now if there is any other pointers or if i missed something in the architecture.

### CODE REVIEW

**Comment by feschber** [comment]
Very nice, thank you! Unfortunately my MacOS VM nuked itself so I will have to set that up again before I can test this... Sorry about thay

**Comment by rohitsangwan01** [comment]
Works for me

File: src/capture/macos.rs [hunk 1]
```diff
@@ -1,33 +1,512 @@
...
+                    value: h as f64,
+                }));
+            }
+        }
+        _ => (),
+    }
+    Ok(())
+}
+
+fn event_tap_tread(
```
Comment (feschber) [inline src/capture/macos.rs:277]:
```suggestion
fn event_tap_thread(
```

File: src/capture/macos.rs [hunk 2]
```diff
@@ -1,33 +1,512 @@
...
+
+        unsafe {
+            configure_cf_settings()?;
+        }
+
+        log::info!("Enabling CGEvent tap");
+        let event_tap_thread_state = state.clone();
+        let event_tap_notify = notify_tx.clone();
+        thread::spawn(move || {
+            event_tap_tread(
```
Comment (feschber) [inline src/capture/macos.rs:397]:
```suggestion
            event_tap_thread(
```

File: src/capture/macos.rs [hunk 3]
```diff
@@ -1,33 +1,512 @@
...
+
+extern "C" {
+    fn CGEventSourceSetLocalEventsSuppressionInterval(
+        event_source: CGEventSource,
+        seconds: CFTimeInterval,
+    );
+}
+
+unsafe fn configure_cf_settings() -> Result<()> {
+    // When we warp the cursor using CGWarpMouseCursorPosition local events are suppressed for a short time
```
[the remark this answers is not in the document]
Comment (meck) [reply src/capture/macos.rs:515]: Yes, i called i directly in an earlier itteration

File: src/capture/macos.rs [hunk 4]
```diff
@@ -1,33 +1,512 @@
...
+    // to the edge of the screen, the cursor will be hidden but we dont want it to appear in a
+    // random location when we exit the client
+    fn reset_mouse_position(&self, event: &CGEvent) -> Result<()> {
+        if let Some((_, pos)) = self.current_client {
+            let location = event.location();
+            let edge_offset = 1.0;
+
+            // After the cursor is warped the next event will carry the delta from the warp
+            // so only half the delta is needed to move the cursor
+            let delta_y = event.get_double_value_field(EventField::MOUSE_EVENT_DELTA_Y) / 2.0;
```
[the remark this answers is not in the document]
Comment (meck) [reply src/capture/macos.rs:109]: Yes, no synthetic event is produced, however the movement delta is tacked onto the next event. This explains the issue https://stackoverflow.com/questions/40904830/constrain-mouse-movement-to-region-in-os-x-without-jitter/40922095#40922095

File: src/capture/macos.rs [hunk 5]
```diff
@@ -1,33 +1,512 @@
...
+        CFRunLoop::get_current().add_source(&tap_source, kCFRunLoopCommonModes);
+    }
+
+    CFRunLoop::run_current();
+
+    let _ = exit.send(Err(anyhow::anyhow!("tap thread exited")));
+}
+
+pub struct MacOSInputCapture {
+    event_rx: Receiver<(u64, Event)>,
```
[the remark this answers is not in the document]
Comment (meck) [reply src/capture/macos.rs]: fixed

File: src/capture/macos.rs [hunk 6]
```diff
@@ -1,33 +1,512 @@
...
+            }
+        }
+        _ => (),
+    }
+    Ok(())
+}
+
+fn event_tap_tread(
+    client_state: Arc<Mutex<InputCaptureState>>,
+    event_tx: Sender<(u64, Event)>,
```
[the remark this answers is not in the document]
Comment (meck) [reply src/capture/macos.rs]: fixed

File: src/capture/macos.rs [hunk 7]
```diff
@@ -1,33 +1,512 @@
...
+    fn CGEventSourceSetLocalEventsSuppressionInterval(
+        event_source: CGEventSource,
+        seconds: CFTimeInterval,
+    );
+}
+
+unsafe fn configure_cf_settings() -> Result<()> {
+    // When we warp the cursor using CGWarpMouseCursorPosition local events are suppressed for a short time
+    // this leeds to the cursor not flowing when crossing back from a clinet, set this to to 0 stops the warp
+    // from working, set a low value by trail and error, 0.05s seems good. 0.25s is the default
```
[the remark this answers is not in the document]
Comment (meck) [reply src/capture/macos.rs]: fixed

File: src/capture/macos.rs (hunk 1 above)
Comment (meck) [reply src/capture/macos.rs]: fixed

File: src/capture/macos.rs (hunk 2 above)
Comment (meck) [reply src/capture/macos.rs]: fixed

**Comment by feschber** [comment]
And just FYI: I'm still debating on whether or not Modifier events should even be handled "server side" at all.
The only backend that currently reports them is the wayland / layershell backend and similarly the only emulation backend that needs modifier events is the wlroots one.
Still good to have them but they also dont work in Windows and Libei Input Captures (even though libei should report them but apparently does not do so)

**Comment by meck** [comment]
I see, im using Hyprland so wayland/wlroots is what im using. I guess removing the events would mean tracking modifier state in wlroots emulation.

**Comment by feschber** [comment]
> I see, im using Hyprland so wayland/wlroots is what im using. I guess removing the events would mean tracking modifier state in wlroots emulation.

Yeah exactly. Definitely keep it in though! Better to have it working in MacOS now.

**Comment by meck** [comment]
I hope to get back to this soon, haven't had the time, the only thing i wanted to add was detection of monitor layout changes, there is a API in `CGDisplayRegisterReconfigurationCallback` however it needs a `CFRunLoop` to run on the main thread for the callback to actually be called, That would mean a lot of architectural  changes to rest of the project so I'm looking for alternatives.

**Comment by feschber** [comment]
no worries! I just got round to testing this. One thing I'm noticing is that the mouse is behaving weird when dragging:
When I press the mouse down and drag it to the left and then release the mouse button, while the mouse is still moving, the mouse jumps back to its original position.

**Comment by arrowmancer** [comment]
Say we have both a Linux machine that we want connected to our MacOS machine (Apple silicon), are we supposed to build and use the same version (this branch) of Lan Mouse on both the Linux and Mac machine in order to e.g. have the pointer go back and forth, or is it sufficient to build this branch on my Mac?  Because I seem to still have trouble having my cursor return from my mac once it's moved there (from my Linux machine).

Also, is anyone for whom this is working having no issues with e.g. having an ipad as a second monitor via sidecar at the same time Lan Mouse is running?

**Comment by arrowmancer** [comment]
> > Say we have both a Linux machine that we want connected to our MacOS machine (Apple silicon), are we supposed to build and use the same version (this branch) of Lan Mouse on both the Linux and Mac machine in order to e.g. have the pointer go back and forth, or is it sufficient to build this branch on my Mac? Because I seem to still have trouble having my cursor return from my mac once it's moved there (from my Linux machine).
> > Also, is anyone for whom this is working having no issues with e.g. having an ipad as a second monitor via sidecar at the same time Lan Mouse is running?
> 
> You will need this branch on the mac side. On the Linux side, you can use the main branch. Keep in mind this is a work in progress.

The work-in-progress state of this PR is definitely understood, I'm just happy it exists for me to try out!  Although sadly I think I'm currently not able to link up my linux and macos machines for the same reasons as here, i.e. the changes caused in the 6.1.1 update with libei:  https://github.com/feschber/lan-mouse/issues/140

I'd be happy to provide any logs!

**Comment by jmatsushita** [comment]
Hi there, thanks for this PR. Just wanted to report that the basic feature works also here with a mac m1/nix-darwin server and a nixos/gnome/wayland client. However:
 - modifier keys don't work (IIUC they works with wlroots, whereas with this setup, I'm using the `Libei` emulation on the nixos side)
 - the mouse wheel was inverted.

Also FYI, I gave a shot at using nixos/cosmic(-epoch) but that didn't work, I suppose this needs https://github.com/Smithay/smithay/pull/1388

**Comment by b0o** [comment]
This is working very well from Sway -> macOS, thank you!

Only issues are that my scroll direction is also inverted, and sometimes the Command key doesn't work (but is fixable by restarting both daemons). These are both issues on the main branch too, so I don't think it's anything specific to your PR.

**Comment by feschber** [comment]
> This is working very well from Sway -> macOS, thank you!
> 
> Only issues are that my scroll direction is also inverted, and sometimes the Command key doesn't work (but is fixable by restarting both daemons). These are both issues on the main branch too, so I don't think it's anything specific to your PR.

Does the command key not work when pressed on MacOS or on the sway client? This PR is only concerned with inputs sent from macos to other devices.

**Comment by b0o** [comment]
Only on macOS when sent from the Linux device.

**Comment by rohitsangwan01** [comment]
@meck now that the current project structure of LanMouse is changed, can you please update your PR accordingly ?

**Comment by feschber** [comment]
> @meck now that the current project structure of LanMouse is changed, can you please update your PR accordingly ?

I can rebase it this weekend

**Comment by feschber** [comment]
@rohitsangwan01 sorry, could not get to it sooner, but it is now rebased.
I can not currently test this though, so lmk if there are any issues that popped up

**Comment by rohitsangwan01** [comment]
@feschber Thank you for the updates, i tested with Ubuntu <-> Mac, its not stable but worked for the first time

**Comment by rohitsangwan01** [comment]
@feschber any plan to merge this PR, its still atleast better then nothing, and can be improved in follow up Pr's i guess.
Till then maybe you can mark MacOS state as Beta

**Comment by feschber** [comment]
> @feschber any plan to merge this PR, its still atleast better then nothing, and can be improved in follow up Pr's i guess.
> Till then maybe you can mark MacOS state as Beta

@meck what do you think?

**Comment by meck** [comment]
> > @feschber any plan to merge this PR, its still atleast better then nothing, and can be improved in follow up Pr's i guess.
> > Till then maybe you can mark MacOS state as Beta
> 
> @meck what do you think?

Hi, sorry for leaving this a bit behind. I've been swamped both in work and personally.  The macOS side features i wanted to add was detection of resolution changes, there was some issues as i mentioned in earlier comment.
And I haven't tested it with any other client then macos -> hyprland. 

I would like to finish it up, but I'm not sure when next I will have the time, but with those caveats I would be fine merging it as is marked as beta if @feschber is too.

**Pull request merged**